kbmd/plugin.c

```diff
@@ -0,0 +1,1240 @@
+			return (errf("PluginError", ret,
+			    "failed to read cert in slot 0x%s", slotstr));
+		}
+
+		if ((ret = add_pubkey(slotstr, cert, pubkeys)) != ERRF_OK ||
+		    (ret = add_attest(pt, slotstr, cert, attest)) != ERRF_OK)
+			return (ret);
+	}
+
+	return (ret);
+}
+
+static errf_t *
+add_serial(nvlist_t *nvl, const struct piv_token *restrict pt)
+{
+	/*
+	 * Only newer (5.0 and presumably later) yubikeys support
+	 * querying the serial.  If not present, we silently skip
+	 * adding.
+	 */
+	if (!ykpiv_token_has_serial(pt))
+		return (ERRF_OK);
+
+	uint32_t serial = ykpiv_token_serial(pt);
+
+	return (envlist_add_uint32(nvl, "serial", serial));
+}
+
+static errf_t *
+pivtoken_to_json(struct piv_token *restrict pt, const char *restrict pin,
+    char **restrict jsonp)
+{
+	errf_t *ret = ERRF_OK;
+	char *json = NULL;
+	nvlist_t *nvl = NULL, *pubkeys = NULL, *attest = NULL;
+	const char *guid = piv_token_guid_hex(pt);
+
+	/*
+	 * First create an nvlist similar in form to the desired JSON.
+	 * Then nvlist_dump_json() is used to take care of proper formatting
+	 * and escaping of the JSON output.
+	 */
+	if ((ret = envlist_alloc(&nvl)) != ERRF_OK ||
+	    (ret = envlist_alloc(&pubkeys)) != ERRF_OK)
+		goto done;
+
+	if ((ret = envlist_add_string(nvl, "guid", guid)) != ERRF_OK ||
+	    (ret = add_cn_uuid(nvl)) != ERRF_OK ||
+	    (ret = envlist_add_string(nvl, "pin", pin)) != ERRF_OK)
+		goto done;
+
+	if ((ret = piv_txn_begin(pt)) != ERRF_OK ||
+	    (ret = piv_select(pt)) != ERRF_OK)
+		goto done;
+
+	if (piv_token_is_ykpiv(pt)) {
+		/*
+		 * Only allocate attest if we're a yubikey (and thus
+		 * support attestation).  Otherwise leave it NULL so
+		 * add_attest() will skip attempting to perform attestation.
+		 */
+		if ((ret = envlist_alloc(&attest)) != ERRF_OK)
+			goto done;
+
+		/*
+		 * The reader on a Yubikey is a part of the Yubikey (as
+		 * opposed to a separate reader + javacard), so we can
+		 * safely assume the reader name is the model of a Yubikey.
+		 */
+		if ((ret = envlist_add_string(nvl, "model",
+		    piv_token_rdrname(pt))) != ERRF_OK ||
+		    (ret = add_serial(nvl, pt)) != ERRF_OK)
+			goto done;
+	}
+
+	if ((ret = add_keys(pt, pubkeys, attest)) != ERRF_OK ||
+	    (ret = envlist_add_nvlist(nvl, "pubkeys", pubkeys)) != ERRF_OK ||
+	    ((attest != NULL &&
+	    (ret = envlist_add_nvlist(nvl, "attestation", attest)) != ERRF_OK)))
+		goto done;
+
+	if ((ret = envlist_dump_json(nvl, &json)) != ERRF_OK)
+		goto done;
+
+	*jsonp = json;
+	json = NULL;
+
+done:
+	if (piv_token_in_txn(pt))
+		piv_txn_end(pt);
+
+	nvlist_free(attest);
+	nvlist_free(pubkeys);
+	nvlist_free(nvl);
+	if (json != NULL)
+		freezero(json, strlen(json) + 1);
+	return (ret);
+}
+
+static errf_t *
+plugin_pivtoken_common(struct piv_token *restrict pt, const char *restrict pin,
+    const char *cmd, char *const *args, custr_t *restrict in,
+    custr_t **restrict outp)
+{
+	errf_t *ret = ERRF_OK;
+	custr_t *data[3] = { in, NULL, NULL };
```

Comment:
When we were still working through the plugin bits, I wasn't sure if something additional would be needed, but since it doesn't, I'll go ahead and do that.

Reply:
confirmed